### ISSUE TITLE
How can we support text_to_instance running in parallel?

### ISSUE THREAD

**Issue by wlhgtc** [allenai/allennlp#5050]
Though we have a multi-process data-loader, we use for loops in [predictor](https://github.com/allenai/allennlp/blob/main/allennlp/predictors/predictor.py#L299).

It will make the process hard and slow when we have many inputs( especially when this predictor as a server).

I mean can we use or add some method to make this call(`text_to_instance`) support multi-process ?

@epwalsh  @dirkgr maybe you could help ?

**Comment by dirkgr** [comment]
`text_to_instance()` is not well typed. Different readers us a different signature for that method. It is not a pretty part of the API.

Right now, the way to achieve what you want to do is to override parts of the predictor to call `text_to_instance()` with batches, and make sure that `text_to_instance()` itself supports batches of course. I think the vision readers do this?

**Comment by wlhgtc** [comment]
@dirkgr  Thanks for your advice, but all I found in vision readers is a for [loop](https://github.com/allenai/allennlp-models/blob/main/allennlp_models/vision/dataset_readers/visual_entailment.py#L69-L74)
Here are my demo code and any better solution:
```python
pool = multiprocessing.Pool(20)
pool.map(self.text_to_instance, json_dicts)
```

**Comment by dirkgr** [comment]
If you're on Linux or Mac, try https://github.com/allenai/bettermap instead of `Pool`.

If that works for you, please report back. I have been wondering for a while if we should make bettermap a more official part of our offering. Unfortunately there is no way to make it work on Windows.

**Comment by dirkgr** [comment]
The other solution is to change `text_to_instance()` so it can accept multiple inputs in addition to just one. That does not solve the problem by itself, but it gives you options to do fast things inside of the function.

**Comment by wlhgtc** [comment]
@dirkgr  Thanks for your advice, I will try it later！

**Comment by wlhgtc** [comment]
@dirkgr 
I hav tried bettermap, but it only works when normal call, like
```python
    def _batch_json_to_instances(self, json_dicts: List[JsonDict]) -> List[Instance]:
        map_per_process(self._dataset_reader.text_to_instance, json_dicts)
        return res
    predict_object = Predictor.from_path(...)
    res = predict_object._batch_json_to_instances(records)
```
But if I use this predictor in parallel client, it would crash.

**Comment by wlhgtc** [comment]
And @dirkgr  a more wired thing:

```python
# 1 ms code
tokens_a = self._dataset_reader._tokenizer.tokenize(part_a)
tokens_b = self._dataset_reader._tokenizer.tokenize(part_b)
tokens = self._dataset_reader._tokenizer.add_special_tokens(tokens_a)

# 0.1 ms code
encoded_tokens = self._dataset_reader._tokenizer.tokenizer.encode_plus(tokens_a
                                                                       add_special_tokens=True,
                                                                       return_tensors=None,
                                                                       return_special_tokens_mask=True)
```
This process lower my service, and I wonder can we speed up this process.

**Comment by dirkgr** [comment]
> if I use this predictor in parallel client, it would crash.

How does it crash?

**Comment by dirkgr** [comment]
The first code uses AllenNLP tokens, whereas the second only uses PyTorch tensors. I am not surprised the second approach is faster. It's a tradeoff between speed and flexibility. The AllenNLP `Token` class knows the offsets of the tokens in the source text, it can be used with a `mismatched_embedder`, etc. Tokenizing separately and then adding the special tokens later gives you flexibility to concatenate sequences in different ways, split them if the sequence becomes too long, etc., but if you happen to want to do exactly what `encode_plus` does, then `encode_plus` is faster.

We have not been too bothered by this so far, since we only have to be fast enough to keep the GPU busy during training. There is no point optimizing code that waits on the GPU 90% of the time. But you are doing inference, where the GPU has to do less work, so maybe we aren't keeping the GPU busy in that case? Is that what you're seeing?

**Comment by wlhgtc** [comment]
> The first code uses AllenNLP tokens, whereas the second only uses PyTorch tensors. I am not surprised the second approach is faster. It's a tradeoff between speed and flexibility. The AllenNLP `Token` class knows the offsets of the tokens in the source text, it can be used with a `mismatched_embedder`, etc. Tokenizing separately and then adding the special tokens later gives you flexibility to concatenate sequences in different ways, split them if the sequence becomes too long, etc., but if you happen to want to do exactly what `encode_plus` does, then `encode_plus` is faster.
> 
> We have not been too bothered by this so far, since we only have to be fast enough to keep the GPU busy during training. There is no point optimizing code that waits on the GPU 90% of the time. But you are doing inference, where the GPU has to do less work, so maybe we aren't keeping the GPU busy in that case? Is that what you're seeing?

Yeah, it's about 20% when inference. 
To be honest, the original idea is that : we could support multi `num_workers` to call `text_to_instance`  by `_read` method, which read data from file line by line. Why could we use the same method to support `text_to_instance`  in parallel . The only diff is input is a file or a list of json.

**Comment by dirkgr** [comment]
To solve this properly, and for all models at the same time, we'll need a major re-design of the `Predictor` API. I'm thinking about some of these things, but nothing is ready to code up yet. In the meantime, to fix your issue, you will have to fix the individual predictor. Predictors usually are tied to a specific reader, so you can change the reader to do parallel `text_to_instance`, and then use that from the predictor. It won't fix it for all cases at the same time, but you can do inference quickly yourself, and we'll learn something for the eventual re-design of this feature.

**Comment by dirkgr** [comment]
@wlhgtc, did you get anywhere with this?

**Comment by wlhgtc** [comment]
> @wlhgtc, did you get anywhere with this?

I just replace the original data process (by allennlp) with HF tokenizers, it works on my own service. 
 So I don't do any search :<

**Comment by gaozhiguang** [comment]
when using allennlp-srl for predicting, the utilization rate of GPU is very low and it consumes more CPU, how can i speed up my predicting. i was trying to use multiprocessing, but because of the much consuming of CPU, parallelization reduces the speed.
Here is my code:
srl_predictor = Predictor.from_path("structured-prediction-srl-bert.2020.12.15.tar.gz",cuda_device=gpu_id)
# text_multi is a list, the element in it is a dict, likes {"sentence":sentence}
text_multi = util.move_to_device(text_multi,gpu_id )
text_multi_srl = srl_predictor.predict_batch_json(text_multi)

**Comment by gaozhiguang** [comment]
I just read the source code, util.py，the function move_to_device() can't put my generated string on GPU. How can I process my data into a tensor and put it into the model for predicting?

**Comment by dirkgr** [comment]
@gaozhiguang, do you still need help with this?

**Comment by houliangxue** [comment]
> when using allennlp-srl for predicting, the utilization rate of GPU is very low and it consumes more CPU, how can i speed up my predicting. i was trying to use multiprocessing, but because of the much consuming of CPU, parallelization reduces the speed. Here is my code:
> 
> ```
> srl_predictor = Predictor.from_path("structured-prediction-srl-bert.2020.12.15.tar.gz",cuda_device=gpu_id)
> # text_multi is a list, the element in it is a dict, likes {"sentence":sentence}
> text_multi = util.move_to_device(text_multi,gpu_id )
> text_multi_srl = srl_predictor.predict_batch_json(text_multi)
> ```
请问你的问题解决了吗
